src/peft/tuners/lora/layer.py

```diff
@@ -219,6 +244,76 @@ def dora_init(self, adapter_name: str) -> None:
         dora_layer.update_layer(base_layer=self.get_base_layer(), lora_A=lora_A, lora_B=lora_B, scaling=scaling)
         self.lora_magnitude_vector[adapter_name] = dora_layer
 
+    def _apply_mora(self, x, lora_A, lora_B, scaling, active_adapter):
+        in_f, out_f = self.in_features, self.out_features
+        r = self.r[active_adapter]
+        if active_adapter in self.mora_type:
+            mora_type = self.mora_type[active_adapter]
+        else:
+            mora_type = 1
+
+        if mora_type == 1 or mora_type == 4:
+            sum_inter = in_f // r
+            if in_f % r != 0:
+                pad_size = r - in_f % r
+                x = torch.cat([x, x[..., :pad_size]], dim=-1)
+                sum_inter += 1
+            in_x = x.view(*x.shape[:-1], sum_inter, r).sum(dim=-2)
+        elif mora_type == 2 or mora_type == 3:
+            mr, nr = in_f//r+1, in_f//r
+            m, n = in_f - r*nr, r*mr - in_f
+            mm, nn_ = m*mr, n * nr
+            if m > 0:
+                x_m, x_n = x[..., :mm], x[..., nn_:]
```

Comment:
the original implementation had:
```
x_m, x_n = x[..., :mm], x[..., mm:]
```

but I'm guessing @kongds intended to use nn_ as the index instead?

Reply:
The original implementation should be correct. `mm` is used to split `x` into two parts.
 Detailed explanations can be found https://github.com/kongds/MoRA/issues/7#issuecomment-2143690518.

Reply:
Thanks, so I can drop the `n * nr` part from this?

Reply:
Yes. Thanks for the PR. However,  if merging it into peft, I think it’s better to separate MoRA from LoRA by using new classes like `MoraConfig` and `MoraModel`, since MoRA does not use low-rank matrices in LoRA (for example, the current implementation only uses matrix A in LoRA).